recipes/cmake/3.x.x/conanfile.py

```diff
@@ -55,85 +55,89 @@ def validate(self):
             "Visual Studio": "14",
         }
 
-        compiler = str(self.settings.compiler)
+        compiler = str(self.info.settings.compiler)
         if compiler not in minimal_version:
-            self.output.warn(
+            self.output.warning(
                 "{} recipe lacks information about the {} compiler standard version support".format(self.name, compiler))
-            self.output.warn(
+            self.output.warning(
                 "{} requires a compiler that supports at least C++{}".format(self.name, minimal_cpp_standard))
             return
 
-        version = Version(self.settings.compiler.version)
+        version = Version(self.info.settings.compiler.version)
         if version < minimal_version[compiler]:
             raise ConanInvalidConfiguration(
                 "{} requires a compiler that supports at least C++{}".format(self.name, minimal_cpp_standard))
 
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
     def source(self):
-        get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
 
-    def _configure_cmake(self):
-        if not self._cmake:
-            self._cmake = CMake(self)
+    def generate(self):
+        if self.info.options.bootstrap:
+            tc = AutotoolsToolchain(self)
+            tc.generate()
+        else:
+            tc = CMakeToolchain(self)
             if not self.settings.compiler.cppstd:
-                self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
-            self._cmake.definitions["CMAKE_BOOTSTRAP"] = False
+                tc.variables["CMAKE_CXX_STANDARD"] = 11
+            tc.variables["CMAKE_BOOTSTRAP"] = False
             if self.settings.os == "Linux":
-                self._cmake.definitions["CMAKE_USE_OPENSSL"] = self.options.with_openssl
+                tc.variables["CMAKE_USE_OPENSSL"] = self.options.with_openssl
                 if self.options.with_openssl:
-                    self._cmake.definitions["OPENSSL_USE_STATIC_LIBS"] = not self.options["openssl"].shared
-            if tools.cross_building(self):
-                self._cmake.definitions["HAVE_POLL_FINE_EXITCODE"] = ''
-                self._cmake.definitions["HAVE_POLL_FINE_EXITCODE__TRYRUN_OUTPUT"] = ''
-            self._cmake.configure(source_folder=self._source_subfolder)
-
-        return self._cmake
+                    openssl = self.dependencies["openssl"]
+                    tc.variables["OPENSSL_USE_STATIC_LIBS"] = not openssl.options.shared
+            if cross_building(self):
+                tc.variables["HAVE_POLL_FINE_EXITCODE"] = ''
+                tc.variables["HAVE_POLL_FINE_EXITCODE__TRYRUN_OUTPUT"] = ''
+            tc.generate()
```

Comment:
Hi @uilianries,

I tried adding this code change but it caused widespread failures. I am unclear on what the need for or benefit of adding CMakeDeps (or AutotoolsDeps) to this recipe. I have confirmed that the build process is linking with the libssl.a file coming from my Conan cache:

> [ 54%] Linking C executable curltest
> /usr/bin/cc  -m64 -w -O3 -DNDEBUG -m64 CMakeFiles/curltest.dir/curltest.c.o -o curltest  lib/libcmcurl.a -ldl /home/marc/.conan2/p/858086a77d04574d/p/lib/libssl.a /home/marc/.conan2/p/858086a77d04574d/p/lib/libcrypto.a -lpthread -ldl /home/marc/.conan2/p/858086a77d04574d/p/lib/libssl.a /home/marc/.conan2/p/858086a77d04574d/p/lib/libcrypto.a -lpthread ../cmnghttp2/libcmnghttp2.a ../cmzlib/libcmzlib.a

Do you agree that adding CMakeDeps and AutomakeDeps is indeed unnecessary and, in fact, detrimental? Thanks

P.S. For reference, the failure looks like:

> -- Configuring incomplete, errors occurred!
> See also "/home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/build/5c09c752508b674ca5cb1f2d327b5a2d582866c8/build/Release/CMakeFiles/CMakeOutput.log".
> See also "/home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/build/5c09c752508b674ca5cb1f2d327b5a2d582866c8/build/Release/CMakeFiles/CMakeError.log".
> cmake/3.20.6: 
> CMake Error at /home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/build/5c09c752508b674ca5cb1f2d327b5a2d582866c8/build/Release/CMakeFiles/CMakeTmp/CMakeLists.txt:16 (add_executable):
>   Target "cmTC_dca64" links to target "openssl::openssl" but the target was
>   not found.  Perhaps a find_package() call is missing for an IMPORTED
>   target, or an ALIAS target is missing?
> 
> 
> CMake Error at Utilities/cmlibarchive/CMakeLists.txt:917 (TRY_COMPILE):
>   Failed to generate test project build system.
> Call Stack (most recent call first):
>   Utilities/cmlibarchive/CMakeLists.txt:1967 (CHECK_CRYPTO)
> 
> 
> cmake/3.20.6: WARN: cmake recipe lacks information about the None compiler standard version support
> cmake/3.20.6: WARN: cmake requires a compiler that supports at least C++11
> cmake/3.20.6: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
> cmake/3.20.6: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
> cmake/3.20.6: ERROR: Package '5c09c752508b674ca5cb1f2d327b5a2d582866c8' build failed
> cmake/3.20.6: WARN: Build folder /home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/build/5c09c752508b674ca5cb1f2d327b5a2d582866c8/build/Release
> ERROR: cmake/3.20.6: Error in build() method, line 117
> 	cmake.configure()
> 	ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/build/5c09c752508b674ca5cb1f2d327b5a2d582866c8/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/package/5c09c752508b674ca5cb1f2d327b5a2d582866c8" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/conan/w/prod/BuildSingleReference/.conan/data/cmake/3.20.6/_/_/build/5c09c752508b674ca5cb1f2d327b5a2d582866c8/src"

Reply:
DO you have the full log? The problem is what cmake find files did it use. CMakeDeps should generated all Findxxx.cmake based on Conan recipes, which provide the expected behavior from Conan side. In case we consume from /share/cmake or any other system path, it may result in unpredictable behavior for other users, that's why CMakeDeps is required.

Reply:
This matches how the older recipe was implemented 🤔 we did not previous go through the `find_package` so it would be odd trying to add that back in